packages/daimo-api/src/contract/ClogMatcher.ts

```diff
@@ -0,0 +1,290 @@
+        ...transferLog,
+        foreignToken: token,
+      };
+      this.txHashToTransfers.set(foreignTransfer.transactionHash, [
+        ...(this.txHashToTransfers.get(foreignTransfer.transactionHash) || []),
+        foreignTransfer,
+      ]);
+    }
+
+    // Load FastCCTP Start events and transform into outbound foreign transfers
+    const startResult = await retryBackoff(
+      `StartFastCctp-logs-query-${from}-${to}`,
+      () =>
+        kdb
+          .selectFrom("index.daimo_fast_cctp")
+          .selectAll()
+          .where("chain_id", "=", "" + chainConfig.chainL2.id)
+          .where((e) => e.between("block_num", "" + from, "" + to))
+          .where("tx_hash", "in", txList)
+          .where("log_name", "=", "Start")
+          .execute()
+    );
+
+    for (const row of startResult) {
+      const fastCctpLog = {
+        blockHash: bytesToHex(row.block_hash, { size: 32 }),
+        blockNumber: BigInt(row.block_num),
+        transactionHash: bytesToHex(row.tx_hash, { size: 32 }),
+        transactionIndex: Number(row.tx_idx),
+        logIndex: Number(row.log_idx),
+        address: getAddress(bytesToHex(row.to_token, { size: 20 })),
+        from: getAddress(
+          bytesToHex(assertNotNull(row.from_addr), { size: 20 })
+        ),
+        to: getAddress(bytesToHex(row.to_addr, { size: 20 })),
+        value: BigInt(row.to_amount),
+      };
+
+      const crossChainToken = this.tokenReg.getToken(
+        fastCctpLog.address,
+        Number(row.to_chain_id) // cross-chain (to chain)
+      );
+      if (crossChainToken == null) continue;
+
+      const crossChainTransfer: ForeignTokenTransfer = {
+        ...fastCctpLog,
+        foreignToken: crossChainToken,
+      };
+      this.txHashToTransfers.set(crossChainTransfer.transactionHash, [
+        ...(this.txHashToTransfers.get(crossChainTransfer.transactionHash) ||
+          []),
+        crossChainTransfer,
+      ]);
+    }
+
+    const elapsedMs = (Date.now() - startTime) | 0;
+    console.log(
+      `[SWAP] loaded ${transferResult.length} bundled transfers and ${startResult.length} Start FastCCTP events in ${elapsedMs}ms`
+    );
+  }
+
+  /**
+   * For inbound cross-chain transfers, we need to fetch the cross-chain event
+   * that initiated the transfer (i.e. the corresponding Start <> FastFinish).
+   */
+  async loadCrossChainClogs(
+    kdb: Kysely<IndexDB>,
+    from: number,
+    to: number,
+    chainId: number,
+    txHashes: Set<Hex>
+  ) {
+    const startTime = Date.now();
+
+    const txList = [...txHashes].map((tx) => Buffer.from(hexToBytes(tx)));
+
+    // Load the Start event on chain A corresponding to the FastCCTP event on
+    // home chain B.
+    const fastCctpRes = await retryBackoff(
+      `clogMatcher-fastCctp-logs-query-${from}-${to}`,
+      () =>
+        kdb
+          .selectFrom("index.daimo_fast_cctp as ff")
+          .leftJoin("index.daimo_fast_cctp as start", (join) =>
```

Comment:
can make this a regular inner join

left join = will return all FastFinish events even when there's no corresponding Start event